railties/lib/rails/tasks/routes.rake

```diff
@@ -1,7 +1,28 @@
+require 'optparse'
+
 desc 'Print out all defined routes in match order, with names. Target specific controller with CONTROLLER=x.'
 task routes: :environment do
   all_routes = Rails.application.routes.routes
   require 'action_dispatch/routing/inspector'
   inspector = ActionDispatch::Routing::RoutesInspector.new(all_routes)
-  puts inspector.format(ActionDispatch::Routing::ConsoleFormatter.new, ENV['CONTROLLER'])
+
+  if ENV['CONTROLLER']
+    ActiveSupport::Deprecation.warn('`CONTROLLER` environment is deprecated and will be removed in Rails 5.1. Please use -g and -c keys instead.')
+  end
+
+  options = { controller: ENV['CONTROLLER'] }
+
+  OptionParser.new do |opts|
+    opts.banner = "Usage: rake routes [options]"
+    opts.on("-c", "--controller [PATTERN]") do |pattern|
+      options[:controller] = pattern
```

Comment:
Perhaps we should move `downcase.sub(/_?controller\z/, '')` to here, and just set `options[:pattern]`. That would simplify the code further down a bit.